WL-243 LMS Language Selection

asadiqbal08 opened the pull request:
We will update the UI or anything else as per suggestion(s). 

<img width="1269" alt="screen shot 2016-01-15 at 5 57 57 pm" src="https://cloud.githubusercontent.com/assets/7334669/12353515/9f2a76fc-bbb1-11e5-9371-f346cd4ef7a7.png">

asadiqbal08 commented on lms/static/js/lang-edx.js:
```diff
@@ -0,0 +1,47 @@
+                            // submit the form in success callback.
+                            $("#settings-form").submit();
+                        },
+                        error: function (xhr, ajaxOptions, thrownError) {
```
To-Be-Implement: Try to show error !

asadiqbal08 commented:
jenkins run all

asadiqbal08 commented on lms/static/js/lang-edx.js:
```diff
@@ -0,0 +1,46 @@
+                            // submit the form in success callback.
+                            $("#settings-form").submit();
+                        },
+                        error: function () {
```
To-Be-Implement: Try to show error !

asadiqbal08 commented:
jenkins run bokchoy

asadiqbal08 commented:
@mattdrayer can you share your thoughts on this PR  ?

mattdrayer commented on lms/djangoapps/ccx/tests/test_field_override_performance.py:
```diff
@@ -214,21 +214,21 @@ class TestFieldOverrideMongoPerformance(FieldOverridePerformanceTestCase):
 
```
@asadiqbal08 can you explain why you modified this CCX test dataset?  It is not clear to me how it relates to the language selection capability.

asadiqbal08 replied:
Actually @mattdrayer , We are hitting the model dark_lang in navigation.html. Due to that we need to increment `1` in db calls.

andy-armstrong replied:
I would have thought that we must already be hitting dark_lang on every request, so is it possible to cache it? We don't want to add an extra database call to every request just to support this feature.

mattdrayer commented on lms/static/js/lang-edx.js:
```diff
@@ -0,0 +1,47 @@
+             * Listener on changing language from selector.
+             * Send an ajax request to save user language preferences.
+             */
+            listenForChange: function() {
```
Can we be more specific about the type of change we are listening for here?  Perhaps listenForLanguagePreferenceChange or something along those lines.

mattdrayer commented on lms/templates/navigation.html:
```diff
@@ -168,3 +185,6 @@ <h2 class="course-header"><span class="provider">${course.display_org_with_defau
 <%include file="help_modal.html"/>
+<%block name="js_extra">
+<%static:js group='language_edx'/>
+</%block>
```
Missing newline

mattdrayer commented on lms/templates/navigation.html:
```diff
@@ -78,6 +79,22 @@ <h2 class="course-header"><span class="provider">${course.display_org_with_defau
     <ol class="user">
       <li class="primary">
+          <form action="/i18n/setlang/" method="post" class="settings-language-form" id="settings-form">
+            <input  type="hidden" id="csrf_token" name="csrfmiddlewaretoken" value="${csrf_token}">
```
Remove extra space in tag

mattdrayer commented on lms/templates/navigation.html:
```diff
@@ -78,6 +79,22 @@ <h2 class="course-header"><span class="provider">${course.display_org_with_defau
     </ol>
     <ol class="user">
       <li class="primary">
+          <form action="/i18n/setlang/" method="post" class="settings-language-form" id="settings-form">
```
@cptvitamin can you take a quick look at this small HTML chunk?  Thanks!

mattdrayer commented:
@asadiqbal08 so far, so good -- please see the scenarios outlined in the discussion thread for our WL-243 JIRA ticket and ensure we are covering these use cases for language detection and subsequent preference updates.

asadiqbal08 commented on lms/templates/navigation.html:
```diff
@@ -78,6 +79,22 @@ <h2 class="course-header"><span class="provider">${course.display_org_with_defau
+            <input title="preference api" aria-label="language" type="hidden" id="preference-api-url" value="${reverse('preferences_api', kwargs={'username': user.username})}">
+            <label for="settings-language-value"><span class="sr">${_("Choose Language")}</span></label>
+            <select class="input select" id="settings-language-value" name="language">
+              % for language in released_languages():
```
release_languages() fetching languages from `dark_lang_darklangconfig`

asadiqbal08 commented:
@ziafazal It would be great if you also share your thoughts here ? thanks

cptvitamin commented on lms/templates/navigation.html:
```diff
@@ -78,6 +79,22 @@ <h2 class="course-header"><span class="provider">${course.display_org_with_defau
       <li class="primary">
+          <form action="/i18n/setlang/" method="post" class="settings-language-form" id="settings-form">
+            <input type="hidden" id="csrf_token" name="csrfmiddlewaretoken" value="${csrf_token}">
+            <input title="preference api" aria-label="language" type="hidden" id="preference-api-url" value="${reverse('preferences_api', kwargs={'username': user.username})}">
```
@mattdrayer we don't want to put an aria-label on a hidden input. It will force it into some accessibility APIs

cptvitamin commented on lms/templates/navigation.html:
```diff
@@ -78,6 +79,22 @@ <h2 class="course-header"><span class="provider">${course.display_org_with_defau
+          <form action="/i18n/setlang/" method="post" class="settings-language-form" id="settings-form">
+            <input type="hidden" id="csrf_token" name="csrfmiddlewaretoken" value="${csrf_token}">
+            <input title="preference api" aria-label="language" type="hidden" id="preference-api-url" value="${reverse('preferences_api', kwargs={'username': user.username})}">
+            <label for="settings-language-value"><span class="sr">${_("Choose Language")}</span></label>
```
You don't have to worry about the ```for``` attribute or ```id``` attribute on the ```select``` if you use an implied ```label``` (put the closing ```</label>``` element after the closing```</select>``` .  It's not critical, but its less code and less likely to suffer a regression if code changes in the future.

mattdrayer replied:
I'd prefer to keep the ```id``` attribute on this select element because it's helpful with DOM work as well.

mattdrayer commented:
jenkins run lettuce

asadiqbal08 commented:
jenkins run lettuce

ziafazal commented on common/djangoapps/lang_pref/middleware.py:
```diff
@@ -26,3 +28,19 @@ def process_request(self, request):
                 request.session[LANGUAGE_SESSION_KEY] = user_pref
+            else:
+                # Set browser language if it is supported by system.
+                accept = request.META.get('HTTP_ACCEPT_LANGUAGE', '')
```
Can we rename `accept` to `preferred_language`?

ziafazal commented on common/djangoapps/lang_pref/middleware.py:
```diff
@@ -26,3 +28,19 @@ def process_request(self, request):
+                # Set browser language if it is supported by system.
+                accept = request.META.get('HTTP_ACCEPT_LANGUAGE', '')
+                lang_headers = [seq[0] for seq in parse_accept_lang_header(accept)]
+                if len(lang_headers) > 0:
```
We can simplify this statement to ` if lang_headers:`

ziafazal commented on common/djangoapps/lang_pref/middleware.py:
```diff
@@ -26,3 +28,19 @@ def process_request(self, request):
+                        set_user_preference(request.user, LANGUAGE_KEY, browser_lang)
+                        request.session[LANGUAGE_SESSION_KEY] = unicode(browser_lang)
+
+    def _get_released_languages(self):
```
what is the reason for writing this function? it is only used once.

ziafazal commented on common/static/js/lang-edx.js:
```diff
@@ -0,0 +1,65 @@
+                        callback();
+                    },
+                    error: function () {
+                        this.refresh();
```
why we are refreshing the page on `error`?

asadiqbal08 replied:
in case of error we resetting the state of selector/dropdown If api returns any error when setting the user language preferences.

andy-armstrong replied:
IMO this is a confusing way to handle the error. Can we catch it and show a message to the user? The most likely error is that the user's session has timed out, so in that situation refreshing will show the login form. I suppose that isn't too bad although it might seem jarring.

If we do think this is the best way to handle the situation then please add a comment explaining the rationale.

asadiqbal08 replied:
@andy-armstrong , Actually I have introduced a language selector at top of the page as you can see in the screenshot. It seems, difficult for me to decide where should I render a error message. Actually page (form) is also submitting on success SO in case of error I was refreshing the state of the selector to a stable value from database.
Can you tell in which area of UI should we render an error string ?

andy-armstrong replied:
@asadiqbal08 I think you are right that it would be hard to show an error. I recommend leaving this as is but adding a comment to say why you are doing this.

ziafazal commented on common/static/js/lang-edx.js:
```diff
@@ -0,0 +1,65 @@
+        };
+    })(jQuery);
+
+edx.language.init = Language.init();
```
I think we should not add language to global `edx` object if we are don't need it.

asadiqbal08 commented:
@ziafazal I have updated the PR. thanks

mattdrayer commented:
@andy-armstrong FYI on this PR

mattdrayer commented:
@asadiqbal08 -- I caught up with @andy-armstrong today -- he wants to ensure that we are reusing the selector from the profile settings page for this new control.  We will likely need to push that code to the 'openedx' folder of edx-platform in order to properly position the code for reuse.  And of course, you can feel free to enhance the control as well if it does not fit your needs.

andy-armstrong commented:
@symbolist I'd originally suggested that this PR could share logic with the account settings form, but I'm not so sure upon looking at it. It looks like this code is mostly server-side, while the settings logic is client-side. Do you see any opportunities to share code between the two? It would be great to get your thoughts on this.

andy-armstrong commented on common/static/js/lang-edx.js:
```diff
@@ -0,0 +1,65 @@
+                    event.preventDefault();
+
+                    // ajax request to save user language preferences.
+                    self.sendAjaxRequest(preference_api_url, this.value, function () {
```
Our best practice is to not use callbacks, but to use promises instead. ``$.ajax`` returns a promise so it is easy to then chain success and failure handlers to it.

mattdrayer replied:
@asadiqbal08 let's be sure to implement a promise here vs. the callback as @andy-armstrong has recommended, unless doing so is not feasible.

andy-armstrong commented on lms/templates/navigation.html:
```diff
@@ -166,3 +184,6 @@ <h2 class="course-header"><span class="provider">${course.display_org_with_defau
 
 <%include file="help_modal.html"/>
+<%block name="js_extra">
+<%static:js group='language_edx'/>
```
It seems a shame to add another JS file to download to every page. Can we add it to a pre-existing bundle since it is small and being included on almost every page?

asadiqbal08 replied:
@andy-armstrong can you suggest me a location where I can include this JS file ?

andy-armstrong replied:
I think I'd put it into the base_application bundle:

https://github.com/edx/edx-platform/blob/master/lms/envs/common.py#L1239

andy-armstrong commented on lms/djangoapps/ccx/tests/test_field_override_performance.py:
```diff
@@ -214,21 +214,21 @@ class TestFieldOverrideMongoPerformance(FieldOverridePerformanceTestCase):
-        ('ccx', 1, True, False): (48, 6, 13),
-        ('ccx', 2, True, False): (135, 6, 84),
-        ('ccx', 3, True, False): (480, 6, 335),
+        ('no_overrides', 1, True, False): (49, 6, 13),
```
Is there any way to avoid the extra SQL call on every page? Is this to fetch the user's current language preference? I wonder if we are already looking this up to set up the request context, in which case it could be cached or accessed from the context. @ormsbee any suggestions?

andy-armstrong replied:
I see from reading the other comments that this is caused by looking up the dark_lang model. It would be good to remove the extra call if possible and then revert all the changes to these tests.

asadiqbal08 replied:
@andy-armstrong @mattdrayer If you see the `released_languages() common/lang_pref/api.py#16` method then you can see that this method is already caching the `dark_lang` model object So I think It will not effect the performance if we call `released_languages()` method in `navigation.html`, It just fetch the data from model in first hit and It will cache it.

**Now the question is then, why is it required to increment theses values in this above test ?**  If you look into this test above then you will see, test is clearing all system caches before measuring so we are not dependent on cache in this test.
Please correct me if I am missing something.

asadiqbal08 replied:
reverted the changes in test. As check has been gone behind a system configuration condition in navigation.html

andy-armstrong commented on common/djangoapps/lang_pref/middleware.py:
```diff
@@ -26,3 +28,14 @@ def process_request(self, request):
+                    # language with highest priority in browser will be at zero index.
+                    browser_lang = lang_headers[0]
+                    if browser_lang in [seq[0] for seq in released_languages()]:
+                        set_user_preference(request.user, LANGUAGE_KEY, browser_lang)
```
I don't think that we should auto set the user's preference, because their preference may be different than their browser's language for some reason, e.g. if they are logging in at an internet cafe, or a friend's house. A user's preference should only ever be set through an explicit action by the user.

andy-armstrong replied:
Just to be clear, we should obey the browser's chosen languages in the absence of an explicit preference. I just don't think that we should set the user's preference, as it is not their chosen preference but is just the best available option.

andy-armstrong commented on common/static/js/lang-edx.js:
```diff
@@ -0,0 +1,65 @@
+var edx = edx || {},
```
Please add some Jasmine tests for this new JavaScript logic.

andy-armstrong replied:
I'd still like to see Jasmine tests for this logic.

mattdrayer replied:
@asadiqbal08 mentioned to me this morning that he was having some difficulty getting Jasmine to work properly -- the test file was not loading some of the necessary modules after putting it in common/js directory.  I'm not sure of the specifics, however.

andy-armstrong commented on lms/templates/navigation.html:
```diff
@@ -76,6 +77,23 @@ <h2 class="course-header"><span class="provider">${course.display_org_with_defau
     </ol>
     <ol class="user">
       <li class="primary">
+          <form action="/i18n/setlang/" method="post" class="settings-language-form" id="language-settings-form">
```
I'm thinking that there should be a configuration setting to specify whether or not to show the language selector. Is this a feature that we'll be enabling on edx.org? Even if it is, it isn't clear that all Open edX sites will want this.

I also would only show the selector if there is more than one released language. Having a selector with only a single item is unnecessary.

asadiqbal08 replied:
agreed

andy-armstrong commented on common/djangoapps/lang_pref/middleware.py:
```diff
@@ -26,3 +28,14 @@ def process_request(self, request):
+                if lang_headers:
+                    # lang_header is list of tuple
+                    # language with highest priority in browser will be at zero index.
+                    browser_lang = lang_headers[0]
```
I think this logic should loop through all of the browser's language preferences, and match the first one that is available. For example, a user might prefer Swiss German, then regular German, and then finally English. This logic would fail to match German even if we have translations, because we don't have Swiss German.

andy-armstrong commented:
@asadiqbal08 this is looking good. I've finished my first pass through this PR and have a few comments, so let me know when to re-review.

asadiqbal08 commented:
@andy-armstrong @mattdrayer with reference to your comment:
 `this PR could share logic with the account settings form...` 
Actually, account settings form is based on backbone logic and it is client side (with in separate layer) and the logic in this PR is not dependent on account setting model. It seems me we could not use selector from account settings page.

asadiqbal08 commented:
jenkins run bokchoy

mattdrayer commented:
Thanks @asadiqbal08 -- @andy-armstrong would you mind taking another look at some point today?

mattdrayer commented on common/djangoapps/lang_pref/middleware.py:
```diff
@@ -26,3 +28,12 @@ def process_request(self, request):
             if user_pref:
                 request.session[LANGUAGE_SESSION_KEY] = user_pref
+            else:
+                # Set browser language if it is supported by system.
```
We are not actually setting the browser language here -- we are setting the session language to the browser language, if it is supported.

mattdrayer commented on common/djangoapps/lang_pref/middleware.py:
```diff
@@ -26,3 +28,12 @@ def process_request(self, request):
+                preferred_language = request.META.get('HTTP_ACCEPT_LANGUAGE', '')
+                lang_headers = [seq[0] for seq in parse_accept_lang_header(preferred_language)]
+                for browser_lang in lang_headers:
+                    if browser_lang in [seq[0] for seq in released_languages()]:
```
It might be better to set a "released_languages" parameter outside of the loop for performance reasons.

mattdrayer commented on lms/templates/navigation.html:
```diff
@@ -75,6 +76,28 @@ <h2 class="course-header"><span class="provider">${course.display_org_with_defau
       </%block>
     </ol>
     <ol class="user">
+        % if settings.FEATURES.get('SHOW_LANGUAGE_SELECTOR', False):
```
It would be nice if we could hide this check behind lang_pref instead of referencing the configuration directly in the template.

mattdrayer replied:
Or perhaps there is a better place to abstract, such as static_content

andy-armstrong commented:
@asadiqbal08 Thanks for addressing all my feedback. :+1: from me.

asadiqbal08 commented:
jenkins run a11y

asadiqbal08 commented:
jenkins run a11y

asadiqbal08 commented:
@mattdrayer I have updated the PR and at-last I find out the issue in loading of Jasmine tests :) 
- I have written some Jasmine tests to support changes. 
- one bokchoy test was failing on last build , I updated it (Let C on Jenkins).
- I had rebased it with master and squash the commits. 
Kindly look into the PR and let me know IF anything else is required or missing. 

Thanks

mattdrayer commented:
@asadiqbal08 I think this PR has gone for long enough -- thank you very much for addressing all of our feedback -- it looks really good to me now and is ready to go IMHO -- :+1:  -- you have merge rights, so you are clear to proceed.

mattdrayer commented:
@pdesjardins FYI

asadiqbal08 commented:
Hurry ... Merging !

symbolist commented:
@andy-armstrong I was away last week so couldn't look at this before. I think [LanguagePreferenceFieldView](https://github.com/edx/edx-platform/blob/master/lms/static/js/student_account/views/account_settings_fields.js#L22) would have done the trick.

mattdrayer commented:
@symbolist, we could log a follow-on story+PR if you think it would be helpful?
